File: problems/subsegment/subsegment.py
#!/usr/bin/env python3
import math
import pprint
import os.path
import sys

def trimWords(words):
   return words.replace(',', '').replace('.','').lower()

def maxCountInArrays(keywordArr, segmentArr):
   num = 0
   for i in range(0, len(keywordArr)-1):
      cnt = 0
      for j in range(0, len(segmentArr)-1):
         if keywordArr[i] == segmentArr[j]:
            cnt += 1
      if cnt > num:
         num = cnt
   return num

def readFileToArray(path):
   f = open(path)
   arr = [line.strip() for line in f] 
   f.close()
   return arr

def permute(arr, num, top, pos):
   quotient = num / top
   remainder = num % top
   arr[pos] = remainder
   if (quotient < top):
      arr[pos+1] = quotient
   if (quotient >= top):
      permute(arr, quotient, top, pos+1)
   return arr


## return row with the lowest value of the last column  
## [7,8,3,5,7]
## [7,8,3,5,2] <--------
## [7,8,3,5,4]

def shortestSubsegment(arr):
   shortest = []
   pos = len(arr[0])-1
   num = sys.maxsize 
   for i in range(0, len(arr)):
      combo = arr[i]
      if num > combo[pos]:
         num = combo[pos]
         shortest = combo
   return shortest


## Written by: Kyle Dinh @ https://github.com/kyledinh/toolkit/tree/master/problems/subsegment 
## keywordArr holds the wanted elements for the sub-segment
## segmentArr is the master segment or source segment

if __name__ == "__main__":
   sourcetextfile = 'input.txt'
   if len(sys.argv) > 1:
      if os.path.isfile(sys.argv[1]):      
         sourcetextfile = sys.argv[1]
      else:   
         print("invalid source file: ", sys.argv[1])
         sys.exit()

   print("input file: ", sourcetextfile)

   lines = readFileToArray(sourcetextfile)
   depth = int(lines[0])
   keywordArr = lines[1:depth+1]
   segmentArr = trimWords(lines[depth+1]).split(' ')
   breath = maxCountInArrays(keywordArr, segmentArr) 

   # Build matrix with the default -1 value
   matrix = [[-1 for i in range(breath)] for i in range(depth)]

   print("depth: ", depth)
   print("breath: ", breath)
   print(len(lines))
   print("src 1: ", segmentArr[0])

   pprint.pprint(matrix)
   pprint.pprint(keywordArr)
   pprint.pprint(segmentArr)

   # Update matrix with index of keyword finds

   for x in range(0, depth):
      y = 0
      for i in range(0, len(segmentArr)):
         if segmentArr[i] == keywordArr[x]:
            print("Found x: " + str(x) + " y: " + str(y) + " @ " + str(i) + "  " + str(segmentArr[i]) + " for " + str(keywordArr[x]))
            matrix[x][y] = i
            y += 1  

   for x in range(0, depth):
      print(" row of matrix for  " + " ".join(str(i) for i in matrix[x]))

   pprint.pprint(matrix)
   permutation = [0]*depth
   pprint.pprint(permutation)

   results = []
   num_possible = int(math.pow(breath, depth))

   print("total possible: " + str(num_possible))

   for x in range(0, num_possible):
      #arr = [-1 for x in range(depth)]
      n = depth+1
      arr = [-1]*n
      combo = permute(permutation, x, breath, 0)

      for i in range(0, depth):
         c = int(combo[i])
         arr[i] = matrix[i][c]

      distance = max(arr[0:depth]) - min(arr[0:depth])
      arr[depth] = distance

      if (-1 not in arr):
         results.append(arr)

   ## Outputing the results

   pprint.pprint(results)
   print("Number of possible combos: " + str(len(results)))

   if len(results) == 0:
      print("NO SUBSEGMENT FOUND")
   else:
      print(shortestSubsegment(results))

File: problems/subsegment/ruby/Subsegment.rb
class Subsegment

   def maxCountInArrays(keywordArr, segmentArr)
      num = 0
      keywordArr.each do |keyword|
         cnt = 0
         segmentArr.each do |segment|
            if keyword == segment
               #print keyword + " : " + segment + "\n"
               cnt = cnt +1
            end
         end
         if cnt > num
            num = cnt
         end
      end
      return num
   end

   def permute(arr, num, top, pos)
      quotient = num / top 
      remainder = num % top 
      arr[pos] = remainder
      if (quotient < top)
         arr[pos+1] = quotient
      end
      if (quotient >= top)
         permute(arr, quotient, top, pos+1)
      end
      return arr   
   end

   def shortestSubsegment(arr)
      shortest = []
      firstArr = arr[0]
      pos = firstArr.length()-1
      num = 9999 # get max int size for ruby 
      arr.each do |combo|
         if num > combo[pos]
            num = combo[pos]
            shortest = combo
         end
      end
      return shortest
   end

   def getResults(filename)
      if !File.file?(filename)  
         return []
      end
      lines = IO.readlines(filename) 
      if lines.empty?
         puts "ERROR reading " + filename
         return []
      end
      depth = lines[0].to_i
      keywordArr = lines.slice(1, depth).map {|x| x.chomp }
      segmentArr = lines[depth+1].gsub(/[^0-9a-zA-Z ]/, '').downcase.split(' ')
      breath = maxCountInArrays(keywordArr, segmentArr)
      num_possible = breath ** depth
      results = [] # catches all the matches 

      puts "segmentArr: " + segmentArr*" "
      puts "keywordsArr: " + keywordArr*" "
      puts "depth: " + depth.to_s
      puts "breath: " + breath.to_s
      puts "possible permutations: " + num_possible.to_s

      # Build matrix, then update with index of keyword finds
      matrix = Array.new(depth) { Array.new(breath) {-1}} #[[-1,-1,-1] x4 ]

      for x in 0..depth-1
         y = 0
         for i in 0..segmentArr.length
            if segmentArr[i] == keywordArr[x]
               puts "Found  " + i.to_s + " " + segmentArr[i]
               matrix[x][y] = i
               y = y +1
            end
         end
      end

      matrix.each {|r|
         r.each {|c|
            print c.to_s + " "
         }
         print "\n"
      }

      permutation = Array.new(depth, 0) # [0,0,0,0] 
      puts "permutation: " + permutation*" "

      for x in 0..num_possible -1
         arr = Array.new(depth+1) { -1 }
         combo = permute(permutation, x, breath, 0)
         for i in 0..depth-1
            c = combo[i]
            arr[i] = matrix[i][c]
         end

         distance =  arr.max - arr.min
         arr[depth] = distance
         puts arr*" "
   
         if !arr.include? -1
            results.push(arr)
         end
      end
      return results
   end

end

############################################## 
## Subsegment, Kyle Dinh 
## https://github.com/kyledinh/toolkit/blob/master/problems/subsegment/
## ruby 1.9.3p362 (2012-12-25 revision 38607) [x86_64-darwin12.4.0]
## usage example: ruby Subsegment alt.txt
############################################## 

ss = Subsegment.new

filename = "input.txt"
results = []

if !ARGV.empty? 
   filename = ARGV[0]
end

if !File.file?(filename)  
   puts "Program can not find your inputfile: " + filename + "."
else  
   results = ss.getResults(filename)
end 

# Final selection for results and display
if results.empty?
   puts "NO SUBSEGMENT FOUND"
else
   puts "Number of possible combos: " + results.length.to_s
   puts ss.shortestSubsegment(results)*" "
end
